### ISSUE TITLE
GitHub Actions

### ISSUE THREAD

**Issue by melroy89** [OpenNFS/OpenNFS#25]
I see you tried to setup Azure Pipelines, but that is replaced by GitHub Actions now a days. Any help needed to setup a pipeline in GH?

Regards,
Melroy vd Berg

**Comment by AmrikSadhra** [comment]
Hey :) I'm quite happy with Azure Pipleines at the moment, produces artifacts across Mac, Linux and Windows with not too much setup, artifacts are available here, though the build badges don't reflect that builds are occurring: https://type2labs.visualstudio.com/OpenNFS/_build/results?buildId=1174&view=results

The individual badges correspond to three older, deprecated and independent pipelines setup through the UI. Now there is a singular pipeline configured with the azure-pipelines.yml in the root of the project

**Comment by melroy89** [comment]
Hi :)

Nice to hear, well in any case it would be rather easy to migrate if you want: https://help.github.com/en/actions/migrating-to-github-actions/migrating-from-azure-pipelines-to-github-actions#migrating-jobs-and-steps

It will at the end better integrated within GitHub all together. See also:
https://help.github.com/en/actions/configuring-and-managing-workflows/configuring-a-workflow
and:
https://help.github.com/en/actions/configuring-and-managing-workflows/managing-a-workflow-run

Microsoft will in the end focus more and more on GitHub and less on Azure DevOps. 

Regarding the "Publish jobs" (eg. for Linux), where could I download the build artifact now?

**Comment by AmrikSadhra** [comment]
Wow, I wasn't aware they were so similar. The integration also looks like something worth pursuing too. Will definitely be something I look into, thanks so much for the heads up! 

There are some missing resources required for launching OpenNFS, and after a large refactor, the only tracks and vehicles currently supported are those from NFS3 -> I'll try to commit some fixes that perform the extraction of these resources automatically/get them into the repo, so it's just a matter of providing a copy of NFS3. Artifacts available here: https://type2labs.visualstudio.com/OpenNFS/_build/results?buildId=1205&view=artifacts&type=publishedArtifacts

**Comment by janisozaur** [comment]
GA is a bit better integrated into GitHub, you don't have to put up with whatever sillyness there are on Azure to get to logs. Both platforms suffer from inexplicable non-liveness of logs and it's not easy to navigate on GA either (the logs occasionally don't load correctly or come delayed or whatever else).

GA and its actions are surprisingly powerful. Setting the whole vcpkg+caching+cmake cross-compilation using MSVC on Windows, the least amenable of all the platforms, takes about 5 actions and is way easier than trying to actually script it.

There are hardly any occasions where the CI lags or slows us down. 6h timeouts for jobs, quite powerful machines (certainly faster than Travis or AppVeyor, can't compare to AP), enough concurrency to check multiple PRs in parallel.

The downsides include the artifact weirdness, where an artifact acts as a "file sink". You can append to it across multiple actions, which has some benefits, but requires switching your brain into this odd mode. The artifacts can only be zips due to this, even though there is some progress on single-file uploads. The zipper used is of Windows style, so it won't have any of those file attributes stuff, which in turn breaks OSX releases. The reported artifact sizes are all over the places and after a few months I have not managed to figure out what do they actually mean (but artifacts can be multiple-GBs in size if you have such need). The artifacts can only be downloaded by logged-in users (this includes viewing builds logs), so if you want to offer CI builds, you have to upload them _somewhere_, possibly a repo with binaries only.

Overall would recommend, 8/10.

Make sure you read https://www.edwardthomson.com/blog/github_actions_advent_calendar.html.

**Comment by AmrikSadhra** [comment]
Wow, don't think I could have asked for a better overview! Thanks a lot, the login-for-artifact issue is annoying but as you've stated, looks like it can be sidestepped with some work. I'll for sure get on the transition when I next have some spare time.

**Comment by AmrikSadhra** [comment]
Hey @danger89, would you be interested in leading the Github actions migration?

**Comment by AmrikSadhra** [comment]
Bumping this, looking for a contributor to lead the migration. Would also like to artifact everything required for ONFS to run instead of just the executable, any help greatly appreciated.

**Comment by melroy89** [comment]
@AmrikSadhra I think I can give support yes. Professionally I have some direct contacts at GitHub that I may even can use if we are really stuck. 

So basically we want to migrate this file:
https://github.com/OpenNFS/OpenNFS/blob/main/azure-pipelines.yml

Towards GitHub Actions. 

I see you try to start: https://github.com/OpenNFS/OpenNFS/runs/1193691077?check_suite_focus=true

**Comment by melroy89** [comment]
I will create a PR with a working build.

**Comment by melroy89** [comment]
I got a working build at least :+1:

**Comment by AmrikSadhra** [comment]
Nice work, very glad this is moving now. Builds on Windows/Mac and the current Linux steps as a matrix would be a nice next step I think?

**Comment by melroy89** [comment]
Yes you are right. Introducing a matrix is the next logical step.
